Handwritten Administrative API client

danielledeleo opened the pull request.

Commit:
Add raw output from `make spec/target/go` from SCN

Generated from commit d78a6bcc3e87f0786693598dda3956828eefe9fa in
suborbital/scn

Commit:
Remove noisy OpenAPI files

Commit:
Add outer Go module

Commit:
Add wrapper scaffolding and a few endpoints

- Runnables are created from a Client + Config

This commit lays out the rough structure of how I think this library
should look and behave.

Commit:
Add execution adapter to wrapper

Commit:
Remove OpenAPI generated libraries

- Add handwritten Administrative API client functions

ospencer commented on admin_requests.go:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	token := TokenResponse{}
+	dec := json.NewDecoder(res.Body)
```
Could we instead use JSON Unmarshal here?

danielledeleo replied:
`json.Unmarshal` takes a `[]byte` slice, whereas `NewDecoder` takes an `io.Reader` interface, which is what `res.Body` is. I'm pretty sure doing it this way pipes the HTTP response output stream directly into the decoder and saves us a memcpy somewhere.

This is a pretty common pattern in Go.

ospencer commented on admin_requests.go:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	if token.Token == "" {
+		return res, errors.Wrapf(errors.New("TokenReponse.Token was empty"),
```
Do we need to check this case? I think it should be caught by the above error check 🤔

danielledeleo replied:
This covers the (highly) unlikely case that the server returns a valid JSON object where the `token` field is explicitly empty. :P

cohix commented on admin_requests.go:
```diff
@@ -0,0 +1,41 @@
+		return nil, err
+	}
+
+	res, err := c.httpClient.Do(req)
```
Let's create a `client.do` function that wraps `httpClient.Do` and also returns an error if the response code is not 2XX, since that wouldn't cause an error normally.

danielledeleo replied:
ok!

ospencer commented on client.go:
```diff
@@ -0,0 +1,56 @@
+	return client, nil
+}
+
+func NewLocalClient() (*Client, error) {
```
I had a bit of a brain fart when reading this—maybe add a comment here saying that it uses URIs for my local installation?

danielledeleo replied:
Yes, I'll add per-function go doc comments throughout.

ospencer commented on client.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+func (c *Client) BuildWith(runnable *Runnable, fn Function) (string, *http.Response, error) {
+	return "", nil, nil
```
What's up with this?

danielledeleo replied:
That's a placeholder! Didn't meant to commit that.

cohix commented on config.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+// LocalConfig generates a DefaultConfig() for localhost
+func LocalConfig() *Config {
```
We should have a `KubernetesConfig` as well :)

cohix commented on runnable.go:
```diff
@@ -0,0 +1,60 @@
+
+	token string
+
+	lock sync.Mutex
```
I don't think we need a lock on this, I highly doubt any highly-concurrent code is going to be doing anything with it :)

Commit:
Add client.do() function to wrap httpClient.Do()

- Remove mutex from Runnable

Commit:
Add more Execution and Administrative routes

MacklinEngineering commented on admin_requests.go:
```diff
@@ -35,7 +35,65 @@ func (c *Client) EditorToken(runnable *Runnable) (*http.Response, error) {
 }
+
+func (c *Client) UserFunctions(customerID string, namespace string) ([]*Runnable, *http.Response, error) {
+	// GET /api/v1/functions/{customerID}/{namespace}
```
delete, or did you want the client to know this is the route?

danielledeleo replied:
Do you mean delete the comment? It's handy just for quick reference as I'm writing it, and it's not a secret. What I'll probably end up doing is make a nice "header" comment for each function. `go doc` reads those and generates docs from them, so it'll probably include that URL.

cohix commented on admin_requests.go:
```diff
@@ -0,0 +1,99 @@
+		Results: []ExecResult{},
+	}
+
+	dec := json.NewDecoder(res.Body)
```
We should make sure all of the `res.Body` uses are getting closed, not sure if `Decode` calls `Close()` or not

danielledeleo replied:
Good question, though if we’re returning an `*http.Response` then wouldn’t we want to leave the response body open?

cohix replied:
Hmm maybe, I'm wondering if we need to return the http.Response to the caller at all...

ospencer replied:
I would vote to not return the response—this is effectively just an abstraction over the API, so I'd abstract that away as well.

cohix commented on client.go:
```diff
@@ -0,0 +1,77 @@
+	httpClient *http.Client
+}
+
+func NewClient(config *Config, httpClient *http.Client) (*Client, error) {
```
I'd say for now we don't need to pass in an http client? Just use `DefaultClient` until we have a use-case that requires a custom one

danielledeleo replied:
Sure

cohix commented on client.go:
```diff
@@ -0,0 +1,77 @@
+	return NewClient(LocalConfig(), http.DefaultClient)
+}
+
+func (c *Client) NewRunnable(environment, customerID, namespace, fnName string) *Runnable {
```
I'm unsure about the `NewRunnable` naming here because it sounds like the client is performing some action to create a runnable somewhere, which is not true. Maybe just `Runnable` ? Or does it even need to be a method on `Client` now that it doesn't hold a reference to that client?

danielledeleo replied:
`Runnable` probably shouldn’t even have this type of constructor then. Would it be reasonable to expect all runnables to be created as a result to an API call? Either UserFunctions or a builder endpoint

cohix replied:
I think the user should be able to easily create an object that is a 'reference' to a Runnable, I just don't want them to get confused and think they're making an API call when they aren't.

cohix commented on execution_requests.go:
```diff
@@ -0,0 +1,48 @@
+		return nil, res, errors.Wrap(err, "failed to Client.Exec")
+	}
+
+	if res.StatusCode != http.StatusOK {
```
With the change to `c.do`, I don't think this will work as expected any more?

danielledeleo replied:
Hmm yeah, I'll have to put the status code check first.

cohix commented on execution_requests.go:
```diff
@@ -0,0 +1,48 @@
+		return nil, res, errors.Wrap(err, "failed to Client.Exec")
+	}
+
+	result, err := ioutil.ReadAll(res.Body)
```
`result` here is JSON containing two keys, we should unmarshal it into a struct

danielledeleo replied:
I thought that was only if Atmo returns an error, otherwise it just returns the output bytes of the function directly.

cohix replied:
Yes sorry I got confused with the builder.Build output

cohix commented on response_types.go:
```diff
@@ -0,0 +1,68 @@
+
+// ToRunnable converts a RunnableResponse type to a Runnable
+func (r RunnableResponse) ToRunnable() *Runnable {
+	match := fqfnRegexp.FindStringSubmatch(r.FQFN)
```
There's an `fqfn` package in the Atmo repo that will do this for you without regex, `fqfn.Parse`

cohix commented on runnable.go:
```diff
@@ -0,0 +1,54 @@
+	"fmt"
+)
+
+type Runnable struct {
```
We may want to steal the `Runnable` struct directly from Atmo? That would ensure that they're identical

Commit:
Replace local Runnable with directive.Runnable

- fix Exec error checking

Commit:
Remove *http.Response from all methods.

Since client.do() creates an error for non-200 responses we don't really
need this anymore.

Commit:
Replace Client.httpClient with http.DefaultClient

Commit:
Add Runnable constructor helper functions

Commit:
Add FunctionExecErrors endpoint

Commit:
Add GET endpoints for Builder API

Commit:
Add Build endpoint 🎉

- remove weird errors.Wrap calls
- add nil checks to calls with *Runnable arguments

ospencer commented on README.md:
```diff
@@ -1,2 +1,13 @@
+## Usage
+> TODO
+## About
+The files within the `compute` directory were automatically generated from Suborbital Compute's OpenAPI specifications. This package wraps those libraries into one that is friendlier to use.
```
Is this true anymore?

Commit:
Add Builder draft endpoints

Commit:
Update tests to use random customerID per run

Commit:
Remove Token test

The hard coded token cannot be easily predicted as it depends on the
compute environment. Routes that depend on the correct Bearer token
being set cover this test anyway.
